src/man/man1/mail.1
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)mail.1	6.6.1 (2.11BSD) 1996/10/22
.\"
.TH MAIL 1 "October 22, 1996"
.UC 4
.SH NAME
mail \- send and receive mail
.SH SYNOPSIS
.B mail
[
.B \-v
]
[
.B \-i
]
[
.B \-n
]
[
.B \-s
subject
]
[
user ...
]
.br
.B mail
[
.B \-v
]
[
.B \-i
]
[
.B \-n
]
.B \-f
[
name
]
.br
.B mail
[
.B \-v
]
[
.B \-i
]
[
.B \-n
]
.B \-u
user
.SH INTRODUCTION
.I Mail
is a intelligent mail processing system, which has
a command syntax reminiscent of
.I ed
with lines replaced by messages.
.PP
The
.B \-v
flag puts mail into verbose mode; the details of
delivery are displayed on the users terminal.
The
.B \-i
flag causes tty interrupt signals to be ignored. This is
particularly useful when using
.I mail
on noisy phone lines.
The
.B \-n
flag inhibits the reading of /etc/Mail.rc.
.PP
.I "Sending mail.\ "
To send a message to one or more people,
.I mail
can be invoked with arguments which are the names of people to 
whom the mail will be sent.  You are then expected to type in 
your message, followed
by an \s-2EOT\s0 (control\-D) at the beginning of a line.
A subject may be specified on the command line by using the
.B \-s
flag. (Only the first argument after the
.B \-s
flag is used as a subject; be careful to quote subjects
containing spaces.)
The section below, labeled
.I "Replying to or originating mail,"
describes some features of
.I mail
available to help you compose your letter.
.PP
.I "Reading mail.\ "
In normal usage
.I mail
is given no arguments and checks your mail out of the
post office, then
prints out a one line header of each message there.
The current message is initially the first message (numbered 1)
and can be printed using the
.B print
command (which can be abbreviated \fBp\fR).
You can move among the messages much as you move between lines in
.IR ed ,
with the commands `+' and `\-' moving backwards and forwards, and
simple numbers.
.PP
.I "Disposing of mail.\ "
After examining a message you can
.B delete
(\fBd\fR)
the message or
.B reply
(\fBr\fR)
to it.
Deletion causes the
.I mail
program to forget about the message.
This is not irreversible; the message can be
.B undeleted
(\fBu\fR)
by giving its number, or the
.I mail
session can be aborted by giving the
.B exit
(\fBx\fR)
command.
Deleted messages will, however, usually disappear never to be seen again.
.PP
.I "Specifying messages.\ "
Commands such as
.B print
and
.B delete
can be given a list of message numbers as arguments to apply
to a number of messages at once.
Thus ``delete 1 2'' deletes messages 1 and 2, while ``delete 1\-5''
deletes messages 1 through 5.
The special name ``*'' addresses all messages, and ``$'' addresses
the last message; thus the command
.B top
which prints the first few lines of a message could be used in
``top *'' to print the first few lines of all messages.
.PP
.I "Replying to or originating mail.\ "
You can use the
.B reply
command to
set up a response to a message, sending it back to the
person who it was from.
Text you then type in, up to an end-of-file,
defines the contents of the message.
While you are composing a message,
.I mail
treats lines beginning with the character `~' specially.
For instance, typing ``~m'' (alone on a line) will place a copy
of the current message into the response right shifting it by a tabstop.
Other escapes will set up subject fields, add and delete recipients
to the message and allow you to escape to an editor to revise the
message or to a shell to run some commands.  (These options
are given in the summary below.)
.PP
.I "Ending a mail processing session.\ "
You can end a
.I mail
session with the
.B quit
(\fBq\fR)
command.
Messages which have been examined go to your
.I mbox
file unless they have been deleted in which case they are discarded.
Unexamined messages go back to the post office.
The
.B \-f
option causes
.I mail
to read in the contents of your
.I mbox
(or the specified file)
for processing; when you
.BR quit ,
.I mail
writes undeleted messages back to this file.
The
.B \-u
flag is a short way of doing
"mail
.B \-f
/usr/spool/mail/user".
.PP
.I "Personal and systemwide distribution lists.\ "
It is also possible to create a personal distribution lists so that,
for instance, you can send mail to ``cohorts'' and have it go
to a group of people.
Such lists can be defined by placing a line like
.IP
alias cohorts bill ozalp jkf mark kridle@ucbcory
.PP
in the file \&.mailrc in your home directory.
The current list of such aliases can be displayed with the
.B alias
.B (a)
command in
.IR mail .
System wide distribution lists can be created by editing
/etc/aliases, see
.IR aliases (5)
and
.IR sendmail (8);
these are kept in a different syntax.
In mail you send, personal aliases will be expanded in mail sent
to others so that they will be able to \fBreply\fR to the recipients.
System wide \fIaliases\fR are not expanded when the mail is sent,
but any reply returned to the machine will have the system wide
alias expanded as all mail goes through
.IR sendmail .
.PP
.I "Network mail (ARPA, UUCP, Berknet)\ "
See
.IR mailaddr(7)
for a description of network addresses.
.PP
.I Mail
has a number of options which can be set in the
.I \&.mailrc
file to alter its behavior; thus ``set askcc'' enables the ``askcc''
feature.  (These options are summarized below.)
.SH SUMMARY
(Adapted from the `Mail Reference Manual')
.PP
Each command is typed on a line by itself, and may take arguments
following the command word.  The command need not be typed in its
entirety \- the first command which matches the typed prefix is used.
For commands which take message lists as arguments, if no message
list is given, then the next message forward which satisfies the
command's requirements is used.  If there are no messages forward of
the current message, the search proceeds backwards, and if there are no
good messages at all,
.I mail
types ``No applicable messages'' and
aborts the command.
.TP 12n
.B \-
Goes to the previous message and prints it out.  If given a numeric
argument
.IR n ,
goes to the
.IR n -th
previous message and prints it.
.TP
.B ?
Prints a brief summary of commands.
.TP
.B !
Executes the \s-2UNIX\s0 shell command which follows.
.TP
.B Print
(\fBP\fR)
Like
.B print
but also prints out ignored header fields. See also
.B print
,
.B ignore
and
.B retain.
.TP
.B Reply
(\fBR\fR)
Reply to originator. Does not reply to other
recipients of the original message.
.TP
.B Type
(\fBT\fR)
Identical to the
.B Print
command.
.TP
.B alias
(\fBa\fR) With no arguments, prints out all currently-defined aliases.  With one
argument, prints out that alias.  With more than one argument, creates
an new or changes an on old alias.
.TP
.B alternates
(\fBalt\fR)
The
.B alternates
command is useful if you have accounts on several machines.
It can be used to inform
.I mail
that the listed addresses are really you. When you
.B reply
to messages,
.I mail
will not send a copy of the message to any of the addresses
listed on the
.I alternates
list. If the
.B alternates
command is given with no argument, the current set of alternate
names is displayed.
.TP
.B chdir
(\fBc\fR) Changes the user's working directory to that specified, if given.  If
no directory is given, then changes to the user's login directory.
.TP
.B copy
(\fBco\fR)
The
.B copy
command does the same thing that
.B save 
does, except that it does not mark the messages it
is used on for deletion when you quit.
.TP
.B delete
(\fBd\fR) Takes a list of messages as argument and marks them all as deleted.
Deleted messages will not be saved in
.IR mbox ,
nor will they be available for most other commands.  
.TP
.B dp
(also \fBdt\fR) Deletes the current message and prints the next message.
If there is no next message, 
.I mail
says ``at EOF.''
.TP
.B edit
(\fBe\fR) Takes a list of messages and points the text editor at each one in
turn.  On return from the editor, the message is read back in.
.TP
.B exit
(\fBex\fR or \fBx\fR) Effects an immediate return to the Shell without
modifying the user's system mailbox, his
.I mbox
file, or his edit file in
.BR \-f .
.TP
.B file
(\fBfi\fR)
The same as
.BR folder .
.TP
.B folders
List the names of the folders in your folder directory.
.TP
.B folder
(\fBfo\fR)
The
.B folder
command switches to a new mail file or folder. With no
arguments, it tells you which file you are currently reading.
If you give it an argument, it will write out changes (such
as deletions) you have made in the current file and read in
the new file. Some special conventions are recognized for
the name. # means the previous file, % means your system
mailbox, %user means user's system mailbox, & means
your \~/mbox file, and +folder means a file in your folder
directory.
.TP
.B from
(\fBf\fR) Takes a list of messages and prints their message headers.
.TP
.B headers
(\fBh\fR) Lists the current range of headers, which is an 18 message group.  If
a ``+'' argument is given, then the next 18 message group is printed, and if
a ``\-'' argument is given, the previous 18 message group is printed.
.TP
.B help
A synonym for ?
.TP
.B hold
(\fBho\fR, also \fBpreserve\fR) Takes a message list and marks each
message therein to be saved in the
user's system mailbox instead of in
.IR mbox .
Does not override the
.B delete
command.
.TP
.B ignore
.B N.B.:
.I Ignore
has been superseded by
.I retain.
.br
Add the list of header fields named to the
.IR "ignored list" .
Header fields in the ignore list are not printed
on your terminal when you print a message. This
command is very handy for suppression of certain machine-generated
header fields. The
.B Type
and
.B Print
commands can be used to print a message in its entirety, including
ignored fields. If
.B ignore
is executed with no arguments, it lists the current set of
ignored fields.
.TP
.B mail
(\fBm\fR) Takes as argument login names and distribution group names and sends
mail to those people.
.TP
.B mbox
Indicate that a list of messages be sent to
.I mbox
in your home directory when you quit. This is the default
action for messages if you do
.I not
have the
.I hold
option set.
.TP
.B next
(\fBn\fR like \fB+\fR or CR) Goes to the next message in sequence and types it.
With an argument list, types the next matching message.
.TP
.B preserve
(\fBpre\fR)
A synonym for
.BR hold .
.TP
.B print
(\fBp\fR)
Takes a message list and types out each message on the user's terminal.
.TP
.B quit
(\fBq\fR) Terminates the session, saving all undeleted, unsaved messages in
the user's
.I mbox
file in his login directory, preserving all messages marked with
.B hold
or
.B preserve
or never referenced
in his system mailbox, and removing all other messages from his system
mailbox.  If new mail has arrived during the session, the message
``You have new mail'' is given.  If given while editing a
mailbox file with the
.B \-f
flag, then the edit file is rewritten.  A return to the Shell is
effected, unless the rewrite of edit file fails, in which case the user
can escape with the
.B exit
command.
.TP
.B reply
(\fBr\fR)
Takes a message list and sends mail to the sender and all
recipients of the specified message.
The default message must not be deleted.
.TP
.B respond
A synonym for
.BR reply .
.TP
.B retain
Add the list of header fields named to the
.IR "retained list" .
Only the header fields in the retain list
are shown on your terminal when you print a message.
All other header fields are suppressed.
The
.B Type
and
.B Print
commands can be used to print a message in its entirety.
If
.B retain
is executed with no arguments, it lists the current set of
retained fields.
.TP
.B save
(\fBs\fR) Takes a message list and a filename and appends each message in
turn to the end of the file.  The filename in quotes, followed by the line
count and character count is echoed on the user's terminal.  
.TP
.B set
(\fBse\fR) With no arguments, prints all variable values.  Otherwise, sets
option.  Arguments are of the form
``option=value''
(no space before or after =) or
``option.''
.TP
.B shell
(\fBsh\fR) Invokes an interactive version of the shell.
.TP
.B size
Takes a message list and prints out the size in characters of each
message.
.TP
.B source
(\fBso\fR)
The
.B source
command reads
.I mail
commands from a file.
.TP
.B top
Takes a message list and prints the top few lines of each.  The number of
lines printed is controlled by the variable
.B toplines
and defaults to five.
.TP
.B type
(\fBt\fR) A synonym for
.BR print .
.TP
.B unalias
Takes a list of names defined by
.B alias
commands and discards the remembered groups of users.  The group names
no longer have any significance.
.TP
.B undelete
(\fBu\fR) Takes a message list and marks each message as
.I not
being deleted.
.TP
.B unread
(\fBU\fR) Takes a message list and marks each message as
.I not
having been read.
.TP
.B unset
Takes a list of option names and discards their remembered values;
the inverse of
.BR set .
.TP
.B visual
(\fBv\fR) Takes a message list and invokes the display editor on each message.
.TP
.B write
(\fBw\fR) Similar to
.BR save ,
except that
.I only
the message body (\fIwithout\fP the header) is saved.
Extremely useful for such tasks as sending and receiving source
program text over the message system.
.TP
.B xit
(\fBx\fR) A synonym for
.BR exit .
.TP
.B z
.I Mail
presents message headers in windowfuls as described under the
.B headers
command. You can move
.IR mail 's
attention forward to the next window with the
.B z
command. Also, you can move to the previous window by using
.BR z\- .
.PP
Here is a summary of the tilde escapes,
which are used when composing messages to perform
special functions.  Tilde escapes are only recognized at the beginning
of lines.  The name
``tilde\ escape''
is somewhat of a misnomer since the actual escape character can be set
by the option
.B escape.
.TP 12n
.BR ~! command
Execute the indicated shell command, then return to the message.
.TP
\fB~b\fR name ...
Add the given names to the list of carbon copy recipients but do not make
the names visible in the Cc: line ("blind" carbon copy).
.TP
\fB~c\fR name ...
Add the given names to the list of carbon copy recipients.
.TP
.B ~d
Read the file ``dead.letter'' from your home directory into the message.
.TP
.B ~e
Invoke the text editor on the message collected so far.  After the
editing session is finished, you may continue appending text to the
message.
.TP
\fB~f\fR messages
Read the named messages into the message being sent.
If no messages are specified, read in the current message.
.TP
.B ~h
Edit the message header fields by typing each one in turn and allowing
the user to append text to the end or modify the field by using the
current terminal erase and kill characters.
.TP
\fB~m\fR messages
Read the named messages into the message being sent, shifted right one
tab.  If no messages are specified, read the current message.
.TP
.B ~p
Print out the message collected so far, prefaced by the message header
fields.
.TP
.B ~q
Abort the message being sent, copying the message to
``dead.letter''
in your home directory if
.B save
is set.
.TP
\fB~r\fR filename
Read the named file into the message.
.TP
\fB~s\fR string
Cause the named string to become the current subject field.
.TP
\fB~t\fR name ...
Add the given names to the direct recipient list.
.TP
.B ~v
Invoke an alternate editor (defined by the VISUAL option) on the
message collected so far.  Usually, the alternate editor will be a
screen editor.  After you quit the editor, you may resume appending
text to the end of your message.
.TP
\fB~w\fR filename
Write the message onto the named file.
.TP
\fB~\||\|\fRcommand
Pipe the message through the command as a filter.  If the command gives
no output or terminates abnormally, retain the original text of the
message.  The command
.IR fmt (1)
is often used as
.I command
to rejustify the message.
.TP
.BR ~~ string
Insert the string of text in the message prefaced by a single ~.  If
you have changed the escape character, then you should double
that character in order to send it.
.PP
Options are controlled via the
.B set
and
.B unset
commands.  Options may be either binary, in which case it is only
significant to see whether they are set or not; or string, in which
case the actual value is of interest.
The binary options include the following:
.TP 15n
.B append
Causes messages saved in
.I mbox
to be appended to the end rather than prepended.
(This is set in /etc/Mail.rc
on 2.11BSD systems.)
.TP
.B ask
Causes
.I mail
to prompt you for the subject of each message you send.  If
you respond with simply a newline, no subject field will be sent.
.TP
.B askcc
Causes you to be prompted for additional carbon copy recipients at the
end of each message.  Responding with a newline indicates your
satisfaction with the current list.
.TP
.B autoprint
Causes the
.B delete
command to behave like
.B dp
\- thus, after deleting a message, the next one will be typed
automatically.
.TP
.B debug
Setting the binary option
.I debug
is the same as specifying
.B \-d
on the command line and causes
.I mail
to output all sorts of information useful for debugging
.IR mail .
.TP
.B dot
The binary option
.I dot
causes
.I mail
to interpret a period alone on a line as the terminator
of a message you are sending.
.TP
.B hold
This option is used to hold messages in the system mailbox
by default.
.TP
.B ignore
Causes interrupt signals from your terminal to be ignored and echoed as
@'s.
.TP
.B ignoreeof
An option related to
.I dot
is
.I ignoreeof
which makes
.I mail
refuse to accept a control-d as the end of a message.
.I Ignoreeof
also applies to
.I mail
command mode.
.TP
.B metoo
Usually, when a group is expanded that contains the sender, the sender
is removed from the expansion.  Setting this option causes the sender
to be included in the group.
.TP
.B nosave
Normally, when you abort a message with two \s-2RUBOUT\s0,
.I mail
copies the partial letter to the file ``dead.letter''
in your home directory. Setting the binary option
.I nosave
prevents this.
.TP
.B Replyall
Reverses the sense of
.I reply
and
.I Reply
commands.
.TP
.B quiet
Suppresses the printing of the version when first invoked.
.TP
.B verbose
Setting the option
.I verbose
is the same as using the
.B \-v
flag on the command line. When mail runs in verbose mode,
the actual delivery of messages is displayed on he users
terminal.
.PP
The following options have string values:
.TP 15n
EDITOR
Pathname of the text editor to use in the
.B edit
command and ~e escape.  If not defined, then a default editor is used.
.TP
PAGER
Pathname of the program to use in the
.B more
command or when
.I crt
variable is set.  A default paginator is used if this option is
not defined.
.TP
SHELL
Pathname of the shell to use in the
.B !
command and the ~! escape.  A default shell is used if this option is
not defined.
.TP
VISUAL
Pathname of the text editor to use in the
.B visual
command and ~v escape.
.TP
.B crt
The valued option
.I crt
is used as a threshold to determine how long a message must
be before
.B PAGER
is used to read it.
.TP
.B escape
If defined, the first character of this option gives the character to
use in the place of ~ to denote escapes.
.TP
.B folder
The name of the directory to use for storing folders of
messages. If this name begins with a `/',
.I mail
considers it to be an absolute pathname; otherwise, the
folder directory is found relative to your home directory.
.TP
.B record
If defined, gives the pathname of the file used to record all outgoing
mail.  If not defined, then outgoing mail is not so saved.
.TP
.B toplines
If defined, gives the number of lines of a message to be printed out
with the
.B top
command; normally, the first five lines are printed.
.SH FILES
.if n .ta 2.5i
.if t .ta 1.8i
/usr/spool/mail/*	post office
.br
~/mbox	your old mail
.br
~/.mailrc	file giving initial mail commands
.br
/tmp/R#	temporary for editor escape
.br

src/cmd/mail.c
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/file.h>

#include <ctype.h>
#include <stdio.h>
#include <pwd.h>
#include <utmp.h>
#include <signal.h>
#include <setjmp.h>
#include <string.h>
#include <sysexits.h>
#include <paths.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

    /* copylet flags */
#define REMOTE      1       /* remote mail, add rmtmsg */
#define ORDINARY    2
#define ZAP     3       /* zap header and trailing empty line */
#define FORWARD     4

#define LSIZE       256
#define MAXLET      300     /* maximum number of letters */
#define MAILMODE    0600        /* mode of created mail */

char    line[LSIZE];
char    resp[LSIZE];
struct let {
    long    adr;
    char    change;
} let[MAXLET];
int nlet    = 0;
char    lfil[50];
long    iop;
char    lettmp[] = "/tmp/maXXXXX";
char    maildir[] = _PATH_MAIL;
char    mailfile[] = _PATH_MAIL "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
char    dead[] = "dead.letter";
char    forwmsg[] = " forwarded\n";
FILE    *tmpf;
FILE    *malf;
char    my_name[60];
int error;
int changed;
int forward;
char    from[] = "From ";
int delex();
int flgf;
int flgp;
int delflg = 1;
int hseqno;
jmp_buf sjbuf;
int rmail;

main(argc, argv)
char **argv;
{
    register int i;
    char *name;
    struct passwd *pwent;

    if (!(name = getlogin()) || !*name || !(pwent = getpwnam(name)) ||
        getuid() != pwent->pw_uid)
        pwent = getpwuid(getuid());
    strncpy(my_name, pwent ? pwent->pw_name : "???", sizeof(my_name)-1);
    if (setjmp(sjbuf))
        done();
    for (i=SIGHUP; i<=SIGTERM; i++)
        setsig(i, delex);
    i = mkstemp(lettmp);
    tmpf = fdopen(i, "r+w");
    if (i < 0 || tmpf == NULL)
        panic("mail: %s: cannot open for writing", lettmp);
    /*
     * This protects against others reading mail from temp file and
     * if we exit, the file will be deleted already.
     */
    unlink(lettmp);
    if (argv[0][0] == 'r')
        rmail++;
    if (argv[0][0] != 'r' &&    /* no favors for rmail*/
       (argc == 1 || argv[1][0] == '-' && !any(argv[1][1], "rhd")))
        printmail(argc, argv);
    else
        bulkmail(argc, argv);
    done();
}

setsig(i, f)
int i;
sig_t f;
{
    if (signal(i, SIG_IGN) != SIG_IGN)
        signal(i, f);
}

any(c, str)
    register int c;
    register char *str;
{

    while (*str)
        if (c == *str++)
            return(1);
    return(0);
}

printmail(argc, argv)
    char **argv;
{
    int flg, i, j, print;
    char *p, *getarg();
    struct stat statb;

    setuid(getuid());
    cat(mailfile, maildir, my_name);
    for (; argc > 1; argv++, argc--) {
        if (argv[1][0] != '-')
            break;
        switch (argv[1][1]) {

        case 'p':
            flgp++;
            /* fall thru... */
        case 'q':
            delflg = 0;
            break;

        case 'f':
            if (argc >= 3) {
                strcpy(mailfile, argv[2]);
                argv++, argc--;
            }
            break;

        case 'b':
            forward = 1;
            break;

        default:
            panic("unknown option %c", argv[1][1]);
            /*NOTREACHED*/
        }
    }
    malf = fopen(mailfile, "r");
    if (malf == NULL) {
        printf("No mail.\n");
        return;
    }
    flock(fileno(malf), LOCK_SH);
    copymt(malf, tmpf);
    fclose(malf);           /* implicit unlock */
    fseek(tmpf, 0L, L_SET);

    changed = 0;
    print = 1;
    for (i = 0; i < nlet; ) {
        j = forward ? i : nlet - i - 1;
        if (setjmp(sjbuf)) {
            print = 0;
        } else {
            if (print)
                copylet(j, stdout, ORDINARY);
            print = 1;
        }
        if (flgp) {
            i++;
            continue;
        }
        setjmp(sjbuf);
        fputs("? ", stdout);
        fflush(stdout);
        if (fgets(resp, LSIZE, stdin) == NULL)
            break;
        switch (resp[0]) {

        default:
            printf("usage\n");
        case '?':
            print = 0;
            printf("q\tquit\n");
            printf("x\texit without changing mail\n");
            printf("p\tprint\n");
            printf("s[file]\tsave (default mbox)\n");
            printf("w[file]\tsame without header\n");
            printf("-\tprint previous\n");
            printf("d\tdelete\n");
            printf("+\tnext (no delete)\n");
            printf("m user\tmail to user\n");
            printf("! cmd\texecute cmd\n");
            break;

        case '+':
        case 'n':
        case '\n':
            i++;
            break;
        case 'x':
            changed = 0;
        case 'q':
            goto donep;
        case 'p':
            break;
        case '^':
        case '-':
            if (--i < 0)
                i = 0;
            break;
        case 'y':
        case 'w':
        case 's':
            flg = 0;
            if (resp[1] != '\n' && resp[1] != ' ') {
                printf("illegal\n");
                flg++;
                print = 0;
                continue;
            }
            if (resp[1] == '\n' || resp[1] == '\0') {
                p = getenv("HOME");
                if (p != 0)
                    cat(resp+1, p, "/mbox");
                else
                    cat(resp+1, "", "mbox");
            }
            for (p = resp+1; (p = getarg(lfil, p)) != NULL; ) {
                malf = fopen(lfil, "a");
                if (malf == NULL) {
                    printf("mail: %s: cannot append\n",
                        lfil);
                    flg++;
                    continue;
                }
                copylet(j, malf, resp[0]=='w'? ZAP: ORDINARY);
                fclose(malf);
            }
            if (flg)
                print = 0;
            else {
                let[j].change = 'd';
                changed++;
                i++;
            }
            break;
        case 'm':
            flg = 0;
            if (resp[1] == '\n' || resp[1] == '\0') {
                i++;
                continue;
            }
            if (resp[1] != ' ') {
                printf("invalid command\n");
                flg++;
                print = 0;
                continue;
            }
            for (p = resp+1; (p = getarg(lfil, p)) != NULL; )
                if (!sendmail(j, lfil, my_name))
                    flg++;
            if (flg)
                print = 0;
            else {
                let[j].change = 'd';
                changed++;
                i++;
            }
            break;
        case '!':
            system(resp+1);
            printf("!\n");
            print = 0;
            break;
        case 'd':
            let[j].change = 'd';
            changed++;
            i++;
            if (resp[1] == 'q')
                goto donep;
            break;
        }
    }
   donep:
    if (changed)
        copyback();
}

/* copy temp or whatever back to /var/mail */
copyback()
{
    register int i, c;
    sigset_t set;
    int fd, new = 0;
    struct stat stbuf;

    sigemptyset(&set);
    sigaddset(&set, SIGINT);
    sigaddset(&set, SIGHUP);
    sigaddset(&set, SIGQUIT);
    (void)sigprocmask(SIG_BLOCK, &set, NULL);
    fd = open(mailfile, O_RDWR | O_CREAT, MAILMODE);
    if (fd >= 0) {
        flock(fd, LOCK_EX);
        malf = fdopen(fd, "r+w");
    }
    if (fd < 0 || malf == NULL)
        panic("can't rewrite %s", lfil);
    fstat(fd, &stbuf);
    if (stbuf.st_size != let[nlet].adr) {   /* new mail has arrived */
        fseek(malf, let[nlet].adr, L_SET);
        fseek(tmpf, let[nlet].adr, L_SET);
        while ((c = getc(malf)) != EOF)
            putc(c, tmpf);
        let[++nlet].adr = stbuf.st_size;
        new = 1;
        fseek(malf, 0L, L_SET);
    }
    ftruncate(fd, 0L);
    for (i = 0; i < nlet; i++)
        if (let[i].change != 'd')
            copylet(i, malf, ORDINARY);
    fclose(malf);       /* implict unlock */
    if (new)
        printf("New mail has arrived.\n");
    (void)sigprocmask(SIG_UNBLOCK, &set, NULL);
}

/* copy mail (f1) to temp (f2) */
copymt(f1, f2)
    FILE *f1, *f2;
{
    long nextadr;

    nlet = nextadr = 0;
    let[0].adr = 0;
    while (fgets(line, LSIZE, f1) != NULL) {
        if (isfrom(line))
            let[nlet++].adr = nextadr;
        nextadr += strlen(line);
        fputs(line, f2);
    }
    let[nlet].adr = nextadr;    /* last plus 1 */
}

copylet(n, f, type)
    FILE *f;
{
    int ch;
    long k;
    char hostname[MAXHOSTNAMELEN];

    fseek(tmpf, let[n].adr, L_SET);
    k = let[n+1].adr - let[n].adr;
    while (k-- > 1 && (ch = getc(tmpf)) != '\n')
        if (type != ZAP)
            putc(ch, f);
    switch (type) {

    case REMOTE:
        gethostname(hostname, sizeof (hostname));
        fprintf(f, " remote from %s\n", hostname);
        break;

    case FORWARD:
        fprintf(f, forwmsg);
        break;

    case ORDINARY:
        putc(ch, f);
        break;

    case ZAP:
        break;

    default:
        panic("Bad letter type %d to copylet.", type);
    }
    while (k-- > 1) {
        ch = getc(tmpf);
        putc(ch, f);
    }
    if (type != ZAP || ch != '\n')
        putc(getc(tmpf), f);
}

isfrom(lp)
register char *lp;
{
    register char *p;

    for (p = from; *p; )
        if (*lp++ != *p++)
            return(0);
    return(1);
}

bulkmail(argc, argv)
char **argv;
{
    char *truename;
    int first;
    register char *cp;
    char *newargv[1000];
    register char **ap;
    register char **vp;
    int dflag;

    dflag = 0;
    delflg = 0;
    if (argc < 1) {
        fprintf(stderr, "puke\n");
        return;
    }
    for (vp = argv, ap = newargv + 1; (*ap = *vp++) != 0; ap++)
        if (ap[0][0] == '-' && ap[0][1] == 'd')
            dflag++;
    if (!dflag) {
        /* give it to sendmail, rah rah! */
        unlink(lettmp);
        ap = newargv+1;
        if (rmail)
            *ap-- = "-s";
        *ap = "-sendmail";
        setuid(getuid());
        execv(_PATH_SENDMAIL, ap);
        perror(_PATH_SENDMAIL);
        exit(EX_UNAVAILABLE);
    }

    truename = 0;
    line[0] = '\0';

    /*
     * When we fall out of this, argv[1] should be first name,
     * argc should be number of names + 1.
     */

    while (argc > 1 && *argv[1] == '-') {
        cp = *++argv;
        argc--;
        switch (cp[1]) {
        case 'r':
            if (argc <= 1)
                usage();
            truename = argv[1];
            fgets(line, LSIZE, stdin);
            if (strncmp("From", line, 4) == 0)
                line[0] = '\0';
            argv++;
            argc--;
            break;

        case 'h':
            if (argc <= 1)
                usage();
            hseqno = atoi(argv[1]);
            argv++;
            argc--;
            break;

        case 'd':
            break;

        default:
            usage();
        }
    }
    if (argc <= 1)
        usage();
    if (truename == 0)
        truename = my_name;
    time(&iop);
    fprintf(tmpf, "%s%s %s", from, truename, ctime(&iop));
    iop = ftell(tmpf);
    flgf = first = 1;
    for (;;) {
        if (first) {
            first = 0;
            if (*line == '\0' && fgets(line, LSIZE, stdin) == NULL)
                break;
        } else {
            if (fgets(line, LSIZE, stdin) == NULL)
                break;
        }
        if (*line == '.' && line[1] == '\n' && isatty(fileno(stdin)))
            break;
        if (isfrom(line))
            putc('>', tmpf);
        fputs(line, tmpf);
        flgf = 0;
    }
    putc('\n', tmpf);
    nlet = 1;
    let[0].adr = 0;
    let[1].adr = ftell(tmpf);
    if (flgf)
        return;
    while (--argc > 0)
        if (!sendmail(0, *++argv, truename))
            error++;
    if (error && safefile(dead)) {
        setuid(getuid());
        malf = fopen(dead, "w");
        if (malf == NULL) {
            printf("mail: cannot open %s\n", dead);
            fclose(tmpf);
            return;
        }
        copylet(0, malf, ZAP);
        fclose(malf);
        printf("Mail saved in %s\n", dead);
    }
    fclose(tmpf);
}

sendrmt(n, name)
char *name;
{
    FILE *rmf, *popen();
    register char *p;
    char rsys[64], cmd[64];
    register pid;
    int sts;

    for (p=rsys; *name!='!'; *p++ = *name++)
        if (*name=='\0')
            return(0);  /* local address, no '!' */
    *p = '\0';
    if (name[1]=='\0') {
        printf("null name\n");
        return(0);
    }
skip:
    if ((pid = fork()) == -1) {
        fprintf(stderr, "mail: can't create proc for remote\n");
        return(0);
    }
    if (pid) {
        while (wait(&sts) != pid) {
            if (wait(&sts)==-1)
                return(0);
        }
        return(!sts);
    }
    setuid(getuid());
    if (any('!', name+1))
        (void)sprintf(cmd, "uux - %s!rmail \\(%s\\)", rsys, name+1);
    else
        (void)sprintf(cmd, "uux - %s!rmail %s", rsys, name+1);
    if ((rmf=popen(cmd, "w")) == NULL)
        exit(1);
    copylet(n, rmf, REMOTE);
    exit(pclose(rmf) != 0);
}

usage()
{
    fprintf(stderr, "Usage: mail [ -f ] people . . .\n");
    error = EX_USAGE;
    done();
}

sendmail(n, name, fromaddr)
    int n;
    char *name, *fromaddr;
{
    char file[256];
    int fd;
    struct passwd *pw;
    char buf[128];
    off_t oldsize;

    if (*name=='!')
        name++;
    if (any('!', name))
        return (sendrmt(n, name));
    if ((pw = getpwnam(name)) == NULL) {
        printf("mail: can't send to %s\n", name);
        return(0);
    }
    cat(file, maildir, name);
    if (!safefile(file))
        return(0);
    fd = open(file, O_WRONLY | O_CREAT | O_EXLOCK, MAILMODE);
    if (fd >= 0)
        malf = fdopen(fd, "a");
    if (fd < 0 || malf == NULL) {
        close(fd);
        printf("mail: %s: cannot append\n", file);
        return(0);
    }
    fchown(fd, pw->pw_uid, pw->pw_gid);
    oldsize = ftell(malf);
    (void)sprintf(buf, "%s@%ld\n", name, oldsize);

    copylet(n, malf, ORDINARY);     /* Try to deliver the message */

    /* If there is any error during the delivery of the message,
     * the mail file may be corrupted (incomplete last line) and
     * any subsequent mail will be apparently lost, since the
     * <NL> before the 'From ' won't be there.  So, restore the
     * file to the pre-delivery size and report an error.
     *
     * fflush does "_flag |= _IOERR" so we don't need to check both the
     # return from fflush and the ferror status.
    */
    (void)fflush(malf);
    if (ferror(malf)) {
        printf("mail: %s: cannot append\n", file);
        ftruncate(fd, oldsize);
        fclose(malf);
        return(0);
    }
    fclose(malf);
    return(1);
}

delex(i)
{
    sigset_t sigt;

    if (i != SIGINT) {
        setsig(i, SIG_DFL);
        sigemptyset(&sigt);
        sigaddset(&sigt, i);
        sigprocmask(SIG_UNBLOCK, &sigt, NULL);
    }
    putc('\n', stderr);
    if (delflg)
        longjmp(sjbuf, 1);
    if (error == 0)
        error = i;
    done();
}

done()
{
    unlink(lettmp);
    exit(error);
}

cat(to, from1, from2)
    char *to, *from1, *from2;
{
    register char *cp, *dp;

    cp = to;
    for (dp = from1; *cp = *dp++; cp++)
        ;
    for (dp = from2; *cp++ = *dp++; )
        ;
}

/* copy p... into s, update p */
char *
getarg(s, p)
    register char *s, *p;
{
    while (*p == ' ' || *p == '\t')
        p++;
    if (*p == '\n' || *p == '\0')
        return(NULL);
    while (*p != ' ' && *p != '\t' && *p != '\n' && *p != '\0')
        *s++ = *p++;
    *s = '\0';
    return(p);
}

safefile(f)
    char *f;
{
    struct stat statb;

    if (lstat(f, &statb) < 0)
        return (1);
    if (statb.st_nlink != 1 || (statb.st_mode & S_IFMT) == S_IFLNK) {
        fprintf(stderr,
            "mail: %s has more than one link or is a symbolic link\n",
            f);
        return (0);
    }
    return (1);
}

panic(msg, a1, a2, a3)
    char *msg;
{
    fprintf(stderr, "mail: ");
    fprintf(stderr, msg, a1, a2, a3);
    fprintf(stderr, "\n");
    done();
}
